packages/core/certora/optimistic-asserter/specs/Asserter_Auxiliary.spec

```diff
@@ -0,0 +1,186 @@
+    calldataarg args;
+    uint256 bond1 = getAssertionBond(ID);
+        f(e, args);
+    uint256 bond2 = getAssertionBond(ID);
+
+    assert bond1 == bond2;
+}
+
+rule whoChanged_assertionCurrency(method f, bytes32 ID)
+filtered{f -> !f.isView && !isMultiCall(f)} {
+    env e;
+    calldataarg args;
+    address currency1 = getAssertionCurrency(ID);
+        f(e, args);
+    address currency2 = getAssertionCurrency(ID);
+
+    assert currency1 == currency2;
+}
+
+/*************************************************
+*                Custom rules                    *
+**************************************************/
+// If calling 'assertTruth' twice, it cannot yield the same assertion ID.
+rule cannotAssertTruthTwiceForSameID() {
+    env e1;
+    env e2;
+    calldataarg args1; 
+    calldataarg args2;
+    bytes32 ID1 = assertTruth(e1, args1);
+    bytes32 ID2 = assertTruth(e2, args2);
+
+    // It's enough to check this rule for a single instance of the currency
+    require getAssertionCurrency(ID1) == testERC20;
+    require getAssertionCurrency(ID2) == testERC20;
+    assert ID1 != ID2;
+}
+
+// This rule makes sure that if we can call assertTruth with some parameters 
+// (without reverting), then we can call it again with the same parameters but
+// just changing the liveness.
+rule assertTruthSucceedsForEveryLiveness(uint64 liveness1, uint64 liveness2) 
+{    
+    env e;
+    bytes claim;
+    address asserter;
+    address callbackRecipient;
+    address escalationManager;
+    address currency = testERC20;
+    uint256 bond;
+    bytes32 identifier;
+    bytes32 domainId;
+
+    require liveness1 != liveness2;
+    // Make sure we don't overflow.
+    require liveness2 + e.block.timestamp <= max_uint64;
+
+    // Assuming no previous call to assertTruth was made:
+    bytes32 ID2 = getId(e,claim,callbackRecipient,escalationManager,liveness2,currency,bond,identifier);
+    require getAssertionAsserter(ID2) == 0;
+    storage initState = lastStorage;
+    
+    // Here we "force" the combination of parameters to succeed.
+    assertTruth(e,claim,asserter,callbackRecipient,escalationManager,
+        liveness1,currency,bond,identifier,domainId);
+
+    // Call again, with a different liveness parameter.
+    assertTruth@withrevert(e,claim,asserter,callbackRecipient,escalationManager,
+        liveness2,currency,bond,identifier,domainId) at initState;
+
+    assert !lastReverted;
+}
+
+// Verified
+rule onlyOneAssertionAtATime(method f, bytes32 assertion, bytes32 other) 
+filtered{f -> !f.isView && !isMultiCall(f)} {
+    env e;
+    calldataarg args;
+    // We distinguish between some assertion and 'another' one.
+    require other != assertion;
+
+    bool settled_before = getAssertionSettled(assertion);
+    bool settledOther_before = getAssertionSettled(other);
+    bool resolution_before = getAssertionSettlementResolution(assertion);
+    bool resolutionOther_before = getAssertionSettlementResolution(other);
+        f(e, args);
```

Comment:
I find these rules to be intellectually interesting (and the techniques are cool), but I have trouble understanding why one would really be concerned about this property (one settlement changing at a time). Can you explain your thinking around these sorts of properties? Are these like more fundamental properties that can be used to build more complex rules?

Reply:
This is an example of a common and general rule we implement in some systems. 
Basically we say that, given a certain instance of a main object in the system, and any function that can change it, we'd expect all other instances to not be changed.
It's important because you want to make sure that all operations in the system isolate a particular object from the others. Had we altered two different objects at the same operation, it would have considered as a bug, or unexpected.
However, if you **do** expect that two objects will be changed at the same operation, this rule will be able to find this case for you, and then you can alter the rule to make sure that the correct function does it.

Reply:
I see, so basically you would do the same if the contract managed accounts for individual users using structs, etc. These accounts would be "independent objects" that should not interfere with each other. Makes sense.